File: go/blob.go
package msgpackblob

import "encoding/hex"

// Blob is a MessagePack BLOB supporting read, mutation (copy-on-write) and JSON.
// The zero value is an empty (invalid) blob.
type Blob struct {
	data []byte
}

// NewBlob constructs a Blob from raw bytes (copies).
func NewBlob(data []byte) Blob {
	return Blob{data: append([]byte(nil), data...)}
}

// blobFromVec takes ownership of data without copying.
func blobFromVec(data []byte) Blob {
	return Blob{data: data}
}

// ── raw access ────────────────────────────────────────────────────────
func (b Blob) Data() []byte { return b.data }
func (b Blob) Size() int    { return len(b.data) }
func (b Blob) Empty() bool  { return len(b.data) == 0 }
func (b Blob) Hex() string  { return hex.EncodeToString(b.data) }

// ── validation ────────────────────────────────────────────────────────
func (b Blob) Valid() bool        { return isValid(b.data, len(b.data)) }
func (b Blob) ErrorPosition() int { return errorPosition(b.data, len(b.data)) }

// ── type inspection ───────────────────────────────────────────────────

// Type returns the type of the root element.
func (b Blob) Type() Type {
	if len(b.data) == 0 {
		return TypeNil
	}
	return getType(b.data, len(b.data), 0)
}

// TypeAt returns the type of the element at path.
func (b Blob) TypeAt(path string) Type {
	n := len(b.data)
	rc, istart, _ := lookup(b.data, n, 0, path)
	if rc != rcOK {
		return TypeNil
	}
	return getType(b.data, n, istart)
}

// TypeStr returns the label of the root element type.
func (b Blob) TypeStr() string { return b.Type().String() }

// TypeStrAt returns the label of the element type at path.
func (b Blob) TypeStrAt(path string) string { return b.TypeAt(path).String() }

// ── extraction ────────────────────────────────────────────────────────

// Extract decodes the element at path (nil Value if not found).
func (b Blob) Extract(path string) Value {
	n := len(b.data)
	rc, istart, iend := lookup(b.data, n, 0, path)
	if rc != rcOK {
		return Nil()
	}
	return decodeElement(b.data, n, istart, iend)
}

// ArrayLength returns the element count of the root container, or -1.
func (b Blob) ArrayLength() int64 {
	if len(b.data) == 0 {
		return -1
	}
	return getContainerCount(b.data, len(b.data), 0)
}

// ArrayLengthAt returns the element count of the container at path, or -1.
func (b Blob) ArrayLengthAt(path string) int64 {
	n := len(b.data)
	rc, istart, _ := lookup(b.data, n, 0, path)
	if rc != rcOK {
		return -1
	}
	return getContainerCount(b.data, n, istart)
}

// ── mutation (copy-on-write) ──────────────────────────────────────────
func (b Blob) apply(path string, value Value, mode int) Blob {
	var nb []byte
	encodeValue(&nb, value)
	rc, out := applyEdit(b.data, len(b.data), path, nb, mode)
	if rc == rcOK {
		return blobFromVec(out)
	}
	return b
}

// Set assigns value at path (creating it where appropriate).
func (b Blob) Set(path string, value Value) Blob { return b.apply(path, value, editSet) }

// SetBlob assigns an existing sub-blob at path (embedded verbatim).
func (b Blob) SetBlob(path string, sub Blob) Blob {
	rc, out := applyEdit(b.data, len(b.data), path, sub.data, editSet)
	if rc == rcOK {
		return blobFromVec(out)
	}
	return b
}

// Insert adds value at path only if it does not already exist.
func (b Blob) Insert(path string, value Value) Blob { return b.apply(path, value, editInsert) }

// Replace overwrites value at path only if it already exists.
func (b Blob) Replace(path string, value Value) Blob { return b.apply(path, value, editReplace) }

// ArrayInsert inserts value before the element at the array index path.
func (b Blob) ArrayInsert(path string, value Value) Blob { return b.apply(path, value, editArrayIns) }

// Remove deletes the element at path.
func (b Blob) Remove(path string) Blob {
	rc, out := applyEdit(b.data, len(b.data), path, nil, editRemove)
	if rc == rcOK {
		return blobFromVec(out)
	}
	return b
}

// Patch applies an RFC 7386 merge patch.
func (b Blob) Patch(merge Blob) Blob {
	rc, out := mergePatch(b.data, len(b.data), 0, merge.data, len(merge.data), 0)
	if rc == rcOK {
		return blobFromVec(out)
	}
	return b
}

// ── JSON conversion ───────────────────────────────────────────────────

// ToJSONBytes returns the byte-exact JSON serialisation (may contain non-UTF-8
// string bytes, identical to the C++ library output).
func (b Blob) ToJSONBytes() []byte {
	if len(b.data) == 0 {
		return []byte("null")
	}
	return toJSONBytes(b.data, len(b.data), false, 0)
}

// ToJSON returns the JSON serialisation as a string.
func (b Blob) ToJSON() string { return string(b.ToJSONBytes()) }

// ToJSONPrettyBytes returns the indented JSON serialisation as bytes.
func (b Blob) ToJSONPrettyBytes(indent int) []byte {
	if len(b.data) == 0 {
		return []byte("null")
	}
	if indent < 0 {
		indent = 0
	}
	if indent > 8 {
		indent = 8
	}
	return toJSONBytes(b.data, len(b.data), true, indent)
}

// ToJSONPretty returns the indented JSON serialisation as a string.
func (b Blob) ToJSONPretty(indent int) string { return string(b.ToJSONPrettyBytes(indent)) }

// FromJSONBytes parses JSON bytes into a msgpack blob.
func FromJSONBytes(json []byte) Blob { return blobFromVec(fromJSON(json)) }

// FromJSON parses a JSON string into a msgpack blob.
func FromJSON(json string) Blob { return blobFromVec(fromJSON([]byte(json))) }

File: python/msgpack_blob/blob.py
"""Blob — an owning byte buffer wrapping a msgpack-encoded value."""

from __future__ import annotations

from typing import Optional, Union

from . import _decode as D
from . import _encode as E
from . import _json as J
from . import _mutate as M
from .value import Type, Value

__all__ = ["Blob"]


class Blob:
    """A msgpack BLOB supporting read, mutation (copy-on-write) and JSON."""

    __slots__ = ("_data",)

    def __init__(self, data: Union[bytes, bytearray, memoryview, None] = b"") -> None:
        self._data = bytes(data) if data is not None else b""

    # ── raw access ────────────────────────────────────────────────────
    def data(self) -> bytes:
        return self._data

    def size(self) -> int:
        return len(self._data)

    def empty(self) -> bool:
        return len(self._data) == 0

    def hex(self) -> str:
        return self._data.hex()

    def __len__(self) -> int:
        return len(self._data)

    def __bytes__(self) -> bytes:
        return self._data

    def __eq__(self, other) -> bool:
        if isinstance(other, Blob):
            return self._data == other._data
        return NotImplemented

    def __hash__(self) -> int:
        return hash(self._data)

    def __repr__(self) -> str:  # pragma: no cover - debug aid
        return f"Blob({self._data.hex()})"

    # ── validation ────────────────────────────────────────────────────
    def valid(self) -> bool:
        return D.is_valid(self._data, len(self._data))

    def error_position(self) -> int:
        return D.error_position(self._data, len(self._data))

    # ── type inspection ───────────────────────────────────────────────
    def type(self, path: Optional[str] = None) -> Type:
        n = len(self._data)
        if path is None:
            return Type.NIL if n == 0 else D.get_type(self._data, n, 0)
        rc, istart, _ = D.lookup(self._data, n, 0, path)
        if rc != D.RC_OK:
            return Type.NIL
        return D.get_type(self._data, n, istart)

    def type_str(self, path: Optional[str] = None) -> str:
        return self.type(path).value

    # ── extraction ────────────────────────────────────────────────────
    def extract(self, path: str) -> Value:
        n = len(self._data)
        rc, istart, iend = D.lookup(self._data, n, 0, path)
        if rc != D.RC_OK:
            return Value.nil()
        return D.decode_element(self._data, n, istart, iend)

    def array_length(self, path: Optional[str] = None) -> int:
        n = len(self._data)
        if path is None:
            return -1 if n == 0 else D.get_container_count(self._data, n, 0)
        rc, istart, _ = D.lookup(self._data, n, 0, path)
        if rc != D.RC_OK:
            return -1
        return D.get_container_count(self._data, n, istart)

    # ── mutation (copy-on-write) ──────────────────────────────────────
    def _apply(self, path: str, value: Value, mode: int) -> "Blob":
        new_bin = bytearray()
        E.encode_value(new_bin, value)
        rc, out = M.apply_edit(self._data, len(self._data), path, bytes(new_bin), mode)
        return Blob(out) if rc == M.RC_OK else self

    def set(self, path: str, value: Union[Value, "Blob"]) -> "Blob":
        if isinstance(value, Blob):
            rc, out = M.apply_edit(self._data, len(self._data), path,
                                   value._data, M.EDIT_SET)
            return Blob(out) if rc == M.RC_OK else self
        return self._apply(path, value, M.EDIT_SET)

    def insert(self, path: str, value: Value) -> "Blob":
        return self._apply(path, value, M.EDIT_INSERT)

    def replace(self, path: str, value: Value) -> "Blob":
        return self._apply(path, value, M.EDIT_REPLACE)

    def array_insert(self, path: str, value: Value) -> "Blob":
        return self._apply(path, value, M.EDIT_ARRAY_INS)

    def remove(self, path: str) -> "Blob":
        rc, out = M.apply_edit(self._data, len(self._data), path, b"", M.EDIT_REMOVE)
        return Blob(out) if rc == M.RC_OK else self

    def patch(self, merge_patch: "Blob") -> "Blob":
        rc, out = M.merge_patch(self._data, len(self._data), 0,
                                merge_patch._data, len(merge_patch._data), 0, 0)
        return Blob(out) if rc == M.RC_OK else self

    # ── JSON conversion ───────────────────────────────────────────────
    def to_json(self) -> str:
        if not self._data:
            return "null"
        return J.to_json(self._data, len(self._data), False, 0)

    def to_json_pretty(self, indent: int = 2) -> str:
        if not self._data:
            return "null"
        indent = max(0, min(8, indent))
        return J.to_json(self._data, len(self._data), True, indent)

    @staticmethod
    def from_json(json: Union[str, bytes, None]) -> "Blob":
        return Blob(J.from_json(json))
